monitor/metrics/window.go

```diff
@@ -28,52 +37,76 @@ func NewWindow(windowCap int) *Window {
 		panic("invalid windowCap")
 	}
 
-	w := make([]*api.Metric, 0, windowCap)
+	w := ring.New(windowCap)
 	return &Window{
-		last:   0,
+		last:   nil,
 		window: w,
 	}
 }
 
 // Add adds a new metric to the window. If the window capacity
 // has been reached, the oldest metric (by the time it was added),
-// will be discarded.
+// will be discarded. Add leaves the cursor on the next spot,
+// which is either empty or the oldest record.
 func (mw *Window) Add(m *api.Metric) {
-	if len(mw.window) < cap(mw.window) {
-		mw.window = append(mw.window, m)
-		mw.last = len(mw.window) - 1
-		return
-	}
+	m.TS = time.Now().UnixNano()
+
+	mw.wMu.Lock()
+	mw.window.Value = m
+	mw.window = mw.window.Next()
+	mw.wMu.Unlock()
+
+	mw.lMu.Lock()
+	mw.last = m
```

Comment:
Can't `mw.window.Value` the same as `mw.last`? 

I mean, if instead of doing `Next()` after inserting, you do it before inserting, you can use `Value` as last value and save a variable and a mutex. (At least at first sight).

Reply:
@hsanjuan addressed